src/task/callback.rs

```diff
@@ -0,0 +1,265 @@
+    _phantom: PhantomData<Spawn>,
+}
+
+impl<Spawn> Runner<Spawn> {
+    /// Creates a new runner with given `max_inplace_spin`.
+    pub fn new(max_inplace_spin: usize) -> Self {
+        Self {
+            max_inplace_spin,
+            ..Default::default()
+        }
+    }
+
+    ///  Sets `max_inplace_spin`.
+    pub fn set_max_inplace_spin(&mut self, max_inplace_spin: usize) {
+        self.max_inplace_spin = max_inplace_spin;
+    }
+}
+
+impl<Spawn> Default for Runner<Spawn> {
+    fn default() -> Self {
+        Runner {
+            max_inplace_spin: 4,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<Spawn> crate::Runner for Runner<Spawn>
+where
+    Spawn: LocalSpawn<Task = Task<Spawn>>,
+{
+    type Task = Task<Spawn>;
+    type Spawn = Spawn;
+
+    fn handle(
+        &mut self,
+        spawn: &mut Spawn,
+        mut task: Task<Spawn>,
+        ctx: &<Self::Spawn as LocalSpawn>::TaskContext,
+    ) -> bool {
+        let mut handle = Handle {
+            spawn,
+            ctx,
+            rerun: false,
+        };
+        match task {
+            Task::Mut(ref mut r) => {
+                let mut tried_times = 0;
+                loop {
+                    r(&mut handle);
+                    if !handle.rerun {
+                        return true;
+                    }
+                    tried_times += 1;
+                    if tried_times == self.max_inplace_spin {
+                        break;
+                    }
+                    handle.rerun = false;
+                }
+            }
+            Task::Once(r) => {
+                r(&mut handle);
+                return true;
+            }
+        }
+        spawn.spawn_ctx(task, ctx);
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{RemoteSpawn, Runner as _};
+
+    use std::sync::mpsc;
+
+    #[derive(Default)]
+    struct MockSpawn {
+        spawn_times: usize,
+    }
+
+    impl LocalSpawn for MockSpawn {
+        type Task = Task<MockSpawn>;
+        type TaskContext = ();
+        type Remote = MockSpawn;
+
+        fn spawn_ctx(&mut self, _t: impl Into<Self::Task>, _ctx: &()) {
+            self.spawn_times += 1;
+        }
+
+        fn remote(&self) -> Self::Remote {
+            unimplemented!()
+        }
+    }
+
+    impl RemoteSpawn for MockSpawn {
+        type Task = Task<MockSpawn>;
+        type SpawnOption = ();
+
+        fn spawn_opt(&self, _t: impl Into<Self::Task>, _opt: &()) {
+            unimplemented!()
+        }
+    }
+
+    #[test]
+    fn test_once() {
+        let mut runner = Runner::default();
+        let mut spawn = MockSpawn::default();
+        let (tx, rx) = mpsc::channel();
+        runner.handle(
+            &mut spawn,
+            Task::new_once(move |_| {
+                tx.send(42).unwrap();
+            }),
+            &(),
+        );
+        assert_eq!(rx.recv().unwrap(), 42);
+    }
+
+    #[test]
+    fn test_mut_no_spin() {
```

Comment:
I think spinning is the reason why `spawn_times` is 0.

Reply:
I should have changed the test name to `test_mut_no_respawn` and `test_mut_respawn`...